### PR TITLE
[#828] Support codes for option questions on the dashboard

### PR DESCRIPTION
**Pull request by muloem**

### CODE REVIEW

File: Dashboard/app/js/lib/controllers/survey-controllers.js [hunk 1]
```diff
@@ -899,11 +899,260 @@ FLOW.questionControl = Ember.ArrayController.create({
...
+        order: c.get('length') + 1,
+        questionId: this.get('questionId'),
+    }));
+  },
+
+  /*
+   *  Persist all the newly added options to the data store.
+   *  Options with empty code and empty text fields are dropped
+   *  from the list.  If they were already persisted in the datastore
+   *  they are deleted
```
Comment (jonase) [inline Dashboard/app/js/lib/controllers/survey-controllers.js]: > If they were already persisted in the datastore they are deleted

What does this mean?

File: Dashboard/app/js/lib/controllers/survey-controllers.js [hunk 2]
```diff
@@ -899,11 +899,260 @@ FLOW.questionControl = Ember.ArrayController.create({
...
+    // reset ordering and persist
+    options.forEach(function (option, index) {
+      var code = option.get('code') && option.get('code').trim();
+      var text = option.get('text') && option.get('text').trim();
+      if (!code) {
+        option.set('code', null); // do not send empty string as code
+      }
+
+      // trimmed whitespace
+      option.set('code', code);
```
Comment (jonase) [inline Dashboard/app/js/lib/controllers/survey-controllers.js]: The `if` statement above is undone here?

File: Dashboard/app/js/lib/controllers/survey-controllers.js (hunk 1 above)
Comment (muloem) [inline Dashboard/app/js/lib/controllers/survey-controllers.js]: The idea is if we have any option in the list with blank text and code, we delete it. But if the blank text and code were from an option that had previously been saved to the datastore, we need to remove it from the datastore as well.  Its a bit of a corner case.

File: Dashboard/app/js/lib/controllers/survey-controllers.js (hunk 2 above)
Comment (muloem) [inline Dashboard/app/js/lib/controllers/survey-controllers.js]: good catch! I'll fix that one.

File: Dashboard/app/js/lib/controllers/survey-controllers.js (hunk 1 above)
Comment (jonase) [inline Dashboard/app/js/lib/controllers/survey-controllers.js]: I see, so this is basically how we can delete previously defined options from the datastore :+1:

**Comment by jonase** [comment]
Code reviewed :+1: Well done!

**Pull request merged**